core/network/src/protocol/sync.rs

```diff
@@ -443,49 +452,41 @@ impl<B: BlockT> ChainSync<B> {
 
 	/// Request syncing for the given block from given set of peers.
 	// This is similar to on_block_announce with unknown parent hash.
-	pub fn sync_fork(&mut self, peers: Vec<PeerId>, hash: &B::Hash, number: NumberFor<B>) ->
-		Vec<(PeerId, BlockRequest<B>)>
-	{
+	pub fn sync_fork(&mut self, peers: Vec<PeerId>, hash: &B::Hash, number: NumberFor<B>) {
 		debug!(target: "sync", "Explicit sync request for block {:?} with {:?}", hash, peers);
-		if self.is_known(&hash) || self.is_already_downloading(&hash) {
+		if self.is_known(&hash) {
 			debug!(target: "sync", "Refusing to sync known hash {:?}", hash);
-			return Vec::default();
+			return;
 		}
 
 		let block_status = self.client.block_status(&BlockId::Number(number - One::one()))
 			.unwrap_or(BlockStatus::Unknown);
 		if block_status == BlockStatus::InChainPruned {
 			trace!(target: "sync", "Refusing to sync ancient block {:?}", hash);
-			return Vec::default();
+			return;
 		}
 
 		self.is_idle = false;
-		let mut requests = Vec::new();
-		for peer_id in peers {
-			let peer = if let Some(peer) = self.peers.get_mut(&peer_id) {
-				peer
-			} else {
-				debug!(target: "sync", "Called sync_fork with a bad peer ID");
-				continue;
-			};
-			if let PeerSyncState::AncestorSearch(_, _) = peer.state {
-				continue;
-			}
-
-			if number > peer.best_number {
-				peer.best_number = number;
-				peer.best_hash = hash.clone();
-			}
+		for peer_id in &peers {
+			if let Some(peer) = self.peers.get_mut(peer_id) {
+				if let PeerSyncState::AncestorSearch(_, _) = peer.state {
+					continue;
+				}
 
-			if number <= self.best_queued_number {
-				if let Some(request) = self.download_unknown_stale(&peer_id, &hash) {
-					requests.push((peer_id, request));
+				if number > peer.best_number {
+					peer.best_number = number;
```

Comment:
What is the invariant that sync is meant to uphold? 

Is this always supposed to be the peer's highest block, or is it the peer's best chain? We do have some non longest-chain fork choice rules (1. follow finality, 2. aurababous), so they are not always the same thing.

Reply:
This is just the highest block currently. The sync tracks the longest fork a bit differently from the shorter forks. E.g. it supports parallel block fetch for the longest chain only. It still fetches all forks though so it should not matter which one is actually "best".  There will be a refactoring to get rid of the "main" chain notion, but this is out of scope of this PR.

Reply:
Well, if it's supposed to be the highest block, then the code here is wrong and I wouldn't want to merge it